### PR TITLE
Adapt to Ubuntu 18.04 / Apply linter to itself / Automatically search for clang-format version

### PR DESCRIPTION
**Pull request by mfehr**
Changes
 * Adapted to Ubuntu 18.04 by automatically search for the correct clang-format executable
 * Apply linter to itself, fix lots of linter errors
 * Add README section about how to disable linter functions from inside code

I added a comment to all real changes to distinguish them from the changes caused by the formatting.

~~@eggerk can you please test this on 16.04?~~ Tested, thanks!
~~@igilitschenski if you still run OSX, could you give this a try?~~ Tested by @ntonci, thanks!

### CODE REVIEW

File: bin/init_linter_git_hooks [hunk 1]
```diff
@@ -24,134 +23,154 @@ from __future__ import print_function
...
+
+import requests
 import yaml
 
 default_cpplint = "modified_cpplint.py"
 
 cpplint_url = ""
 
-clang_format_diff_executable = "clang-format-diff-3.8"
+CLANG_FORMAT_DIFF_EXECUTABLE_VERSIONS = [
```
Comment (mfehr) [inline bin/init_linter_git_hooks]: *

File: bin/init_linter_git_hooks [hunk 2]
```diff
@@ -24,134 +23,154 @@ from __future__ import print_function
...
-                         stdout=subprocess.PIPE, stderr=subprocess.PIPE) == 0
+    """Check if a bash command exists."""
+    return subprocess.call(
+        "type " + cmd,
+        shell=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE) == 0
+
+
+def find_clang_format_executable():
```
Comment (mfehr) [inline bin/init_linter_git_hooks]: *

File: linter.py [hunk 3]
```diff
@@ -3,463 +3,490 @@
...
-  merge_msg_file_path = repo_root + "/.git/MERGE_MSG"
-  return os.path.isfile(merge_msg_file_path)
+    """Check if the current commit is a merge commit."""
+    merge_msg_file_path = repo_root + "/.git/MERGE_MSG"
+    return os.path.isfile(merge_msg_file_path)
 
 
-def run_clang_format(repo_root, staged_files, list_of_changed_staged_files):
-  """Runs clang format on all cpp files staged for commit."""
+def find_clang_format_executable():
```
Comment (mfehr) [inline linter.py]: *

File: bin/init_linter_git_hooks [hunk 4]
```diff
@@ -24,134 +23,154 @@ from __future__ import print_function
...
-    print("Failed to copy default pylint.rc.")
-    exit(1)
+    if not os.path.isfile(script_directory + "/pylint.rc"):
+        print("ERROR: Could not download pylint.rc file!")
+        exit(1)
 
-  if not os.path.isfile(script_directory + "/pylint.rc"):
-    print("ERROR: Could not download pylint.rc file!")
-    exit(1)
+    find_clang_format_executable()
```
Comment (mfehr) [inline bin/init_linter_git_hooks]: *

File: linter.py [hunk 5]
```diff
@@ -3,463 +3,490 @@
...
+    clang_format_path = (
+        "/tmp/" + os.path.basename(os.path.normpath(repo_root)) + "_" +
+        datetime.datetime.now().isoformat() + ".clang.patch")
 
-    print("=" * 80)
-    print("Formatted staged C++ files with clang-format.\n" +
-          "Patch: {}".format(clang_format_path))
-    print("=" * 80)
-  return True
+    clang_format_diff_executable = find_clang_format_executable()
```
Comment (mfehr) [inline linter.py]: *

File: linter.py [hunk 6]
```diff
@@ -3,463 +3,490 @@
...
 import os
-import pylint.lint
 import re
 import subprocess
+
+import pylint.lint
 import yaml
 
-CLANG_FORMAT_DIFF_EXECUTABLE = "clang-format-diff-3.8"
+CLANG_FORMAT_DIFF_EXECUTABLE_VERSIONS = [
```
Comment (mfehr) [inline linter.py]: *

**Comment by eggerk** [comment]
Can you also adapt the readme?

**Comment by mfehr** [comment]
@eggerk done, ptal

File: README.md [hunk 7]
```diff
@@ -73,6 +79,10 @@ IncludeCategories:
 ...
 ```
 
+**Python**
+
+Currently there it is not possible to configure the python formatter on a per-repository basis.
```
Comment (eggerk) [inline README.md]: typo

**Pull request merged**